roxie/roxie.md

```diff
@@ -0,0 +1,155 @@
+including potential crashes when activities are destroyed. Therefore we report errors if the
+counts don’t tally properly at the end of a query – but working out where a call was missed
+is often not trivial. Usually it’s because of an exception thrown from an unexpected place,
+e.g. midway through starting.
+
+Note that row requests from the activities above a splitter may execute on whichever thread
+downstream from the splitter happens to need that particular row first.
+
+The splitter code for tracking whether any of the downstream activities still need a row is a
+bit hairy/inefficient, IIRC. There may be scope to optimize (but I would recommend adding
+some good unit test cases first!)
+
+## How does “I beat you to it” work?
+
+When there are multiple agents fulfilling data on a channel, work is shared among them via a
+hash of the packet header, which is used to determine which agent should work on that packet.
+However, if it doesn’t start working on it within a short period (either because the node is
+down, or because it is too busy on other in-flight requests), then another node may take over. The IBYTI messages are used to indicate that a node has started to work on a packet and
+therefore there is no need for a secondary to take over.
+
+The priority of agents as determined by the packet hash is also used to determine how to
+proceed if an IBYTI is received after starting to work on a request. If the IBYTI is from a lower priority buddy (sub-channel) then it is ignored, if it’s from a higher priority one
+then the processing will be abandoned.
+
+When multicast is enabled, the IBYTI is sent on the same multicast channel as the original
+packet (and care is needed to ignore ones sent by yourself). Otherwise it is sent to all
+buddy IPs.
+
+Nodes keep track of how often they have had to step in for a supposedly higher priority node,
+and reduce their wait time before stepping in each time this happens, so if a node has
+crashed then the buddy nodes will end up taking over without every packet being delayed.
+
+(QUESTION – does this result in the load on the first node after the failed node getting double the load?)
+
+Newer code for cloud systems (where the topology may change dynamically) send the information
+about the buddy nodes in the packet header rather than assuming all nodes already have a
+consistent version of that information. This ensures that all agents are using the same
+assumptions about buddy nodes and their ordering.
+
+## All about index compression
+
+An index is basically a big sorted table of the keyed fields, divided into pages, with an
+index of the last row from each page used to be able to locate pages quickly. The bottom
+level pages (‘leaves’) may also contain payload fields that do not form part of the lookup
+but can be returned with it.
+
+Typical usage within LN Risk tends to lean towards one of two cases:
+-	Many keyed fields with a single “DID” field in the payload
```

Comment:
tab character used, rather than spaces.  Also in other places

Reply:
Will fix